### PR TITLE
Adding gDrive synchronization support

### PR DESCRIPTION
**Pull request by rdamazio**

### CODE REVIEW

File: src/model/storage/gdrive.ts [hunk 1]
```diff
@@ -0,0 +1,481 @@
...
+    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+    if (GoogleDriveStorage.DEBUG_SYNC) console.log('SYNC: Disabling');
+
+    this._stopBackgroundSync();
+
+    // Get rid of the token immediately - worst case, if revoking it fails, we'll still be disconnected.
+    const oldToken = this._token;
+    this._token = undefined;
+    (await this._browserStorage).removeItem(GoogleDriveStorage.TOKEN_STORAGE_KEY);
+    this._stateSubject.next(DriveSyncState.DISCONNECTED);
```
Comment (zyv) [inline src/model/storage/gdrive.ts:176]: I'm not sure if it matters, but don't you want to do this even before you trash the token?

```suggestion
    this._stopBackgroundSync();
    this._stateSubject.next(DriveSyncState.DISCONNECTED);
    
    // Get rid of the token immediately - worst case, if revoking it fails, we'll still be disconnected.
    const oldToken = this._token;
    this._token = undefined;
    (await this._browserStorage).removeItem(GoogleDriveStorage.TOKEN_STORAGE_KEY);
```

**Review by zyv — COMMENTED** [review]
Oh wow, that's a PR and a half. Nice work! 👍 

I'm sorry if my comments are underwhelming, I really don't have much time, and to be confident about what I say, I need to properly grasp your state machine and possible transitions.

I guess the general rule should be to switch/check the state as the first and last item in the methods to make sure nothing breaks even if they happen to be called concurrently and/or incorrectly...

However, I'm not able to represent the invariants and check them in my head at the moment. Maybe some tests would be nice, although it's obviously a lot of work.

I may be able to take another look later if that helps.

File: src/model/storage/gdrive.ts (hunk 1 above)
Comment (zyv) [reply src/model/storage/gdrive.ts:176]: P.S. Yes, exactly the possible/desired transition type of question :-/

File: src/model/proto/checklist.proto [hunk 2]
```diff
@@ -13,6 +13,9 @@ message ChecklistFileMetadata {
   string aircraft_info = 5;
   string manufacturer_info = 6;
   string copyright_info = 7;
+
+  // Last modified time in UNIX seconds since the epoch.
+  int32 modified_time = 8;
```
[the remark this answers is not in the document]
Comment (rdamazio) [reply src/model/proto/checklist.proto:18]: Yes, the epoch is defined as UTC. Clarified.

File: src/model/storage/multipart.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
+const CRLF = '\r\n';
```
[the remark this answers is not in the document]
Comment (rdamazio) [reply src/model/storage/multipart.ts:1]: There is, and I tried it (among a few other solutions) - it *almost* works, except the mime type is not respected by gDrive (not sure if that's a limitation of gDrive API or something in FormData) - it instead detects that the contents are JSON and stores them as application/json, and there's no way to set the content encoding to base64.

Documented this.

File: src/model/storage/gdrive.ts [hunk 4]
```diff
@@ -0,0 +1,481 @@
...
+  private static readonly DEBUG_SYNC = true;
+  private static readonly TOKEN_STORAGE_KEY = 'gdrive_token';
+  private static readonly UPLOAD_API_PATH = '/upload/drive/v3/files';
+  // eslint-disable-next-line no-secrets/no-secrets
+  private static readonly CLIENT_ID = '561910372899-o32ockiiaiv1elinrfvcnfelashd0ctl.apps.googleusercontent.com';
+  private static readonly API_SCOPE = 'https://www.googleapis.com/auth/drive.appdata';
+  private static readonly CHECKLIST_MIME_TYPE = 'application/vnd.damazio.efis-editor.checklist';
+  private static readonly CHECKLIST_EXTENSION = '.checklist';
+  private static readonly LONG_SYNC_INTERVAL_MS = 60000;
+  private static readonly SHORT_SYNC_INTERVAL_MS = 10000;
```
[the remark this answers is not in the document]
Comment (rdamazio) [reply src/model/storage/gdrive.ts:31]: Done

File: src/model/storage/gdrive.ts [hunk 5]
```diff
@@ -0,0 +1,481 @@
...
+        return void 0;
+      });
+
+    return Promise.all([
+      apiLoad,
+      this._browserStorage,
+      firstValueFrom(this._checklistStorage.listChecklistFiles()),
+    ]).then((all: [void, Storage, string[]]) => {
+      // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+      if (GoogleDriveStorage.DEBUG_SYNC) console.log('SYNC: gDrive API initialized');
```
[the remark this answers is not in the document]
Comment (rdamazio) [reply src/model/storage/gdrive.ts:85]: Good point. Even better, I just moved all of them to console.debug which is filtered-out by default but can be enabled in the browser's console.

File: src/model/storage/multipart.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (zyv) [reply src/model/storage/multipart.ts:1]: Oh, that's annoying... Documenting is great, would have given me more context. I understand folks are reporting that using the API directly via fetch works, but I haven't dug deep enough into it. So, it seems like this is the only way if you don't want to use the `form-data` library or any other like it:

https://www.npmjs.com/package/enchanted-form-data

<sigh>

File: src/model/storage/gdrive.ts [hunk 6]
```diff
@@ -0,0 +1,481 @@
...
+      },
+    });
+    client.requestAccessToken();
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  private _handleRequestFailure(reason: gapi.client.Response<any>) {
+    this._retryCount++;
+
+    if (reason.status === 401 || reason.status === 403) {
```
[the remark this answers is not in the document]
Comment (rdamazio) [reply src/model/storage/gdrive.ts:147]: Done

File: src/model/storage/gdrive.ts [hunk 7]
```diff
@@ -0,0 +1,481 @@
...
+      // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+      if (GoogleDriveStorage.DEBUG_SYNC) console.log('SYNC: Refreshing token');
+      this._token = undefined;
+
+      if (this._retryCount < 3) {
```
[the remark this answers is not in the document]
Comment (rdamazio) [reply src/model/storage/gdrive.ts:152]: Done

File: src/model/storage/gdrive.ts (hunk 1 above)
Comment (rdamazio) [reply src/model/storage/gdrive.ts:176]: In practice, it probably doesn't matter, but if you're strict about asynchronous operations, then technically, transitioning to DISCONNECTED before the token is gone could mean that another state transition out of DISCONNECTED starts and the token is still kept/used, or even that it gets revoked halfway through another sync (very unlikely given that the await won't ever block in practice).

File: src/model/storage/gdrive.ts [hunk 8]
```diff
@@ -0,0 +1,481 @@
...
+
+    // List remote checklists.
+    const remoteFiles = this._listFiles().then((existingFiles: gapi.client.drive.File[]) => {
+      // Map from file names to gDrive file IDs.
+      const remoteFileMap = new Map<string, gapi.client.drive.File>();
+      for (const file of existingFiles) {
+        let name = file.name;
+        if (!name || !file.id) continue;
+
+        // Checklist mime type.
```
[the remark this answers is not in the document]
Comment (rdamazio) [reply src/model/storage/gdrive.ts:210]: Google Drive does not have a UI that lets users see or manipulate these files, they're stored as "app data" (hence the appdata scope) - if you go to drive.google.com, Settings, Manage Apps, that's where this shows up, and it only lets you see space usage, not actual files.

That said, API-wise it wasn't obvious that _listFiles was already filtering by MIME type, so made that more explicit.

File: src/model/storage/gdrive.ts [hunk 9]
```diff
@@ -0,0 +1,481 @@
...
+      modifiedTime: mtime.toISOString(),
+      // If file was previously trashed and we're re-uploading it, take it out of the trash.
+      trashed: false,
+    };
+
+    const multipart = new MultipartEncoder();
+    multipart.addPart(JSON.stringify(metadata), { mimeType: 'application/json' });
+    multipart.addPart(contents, { mimeType: mimeType, base64encode: true });
+
+    const uploaded = await gapi.client.request({
```
[the remark this answers is not in the document]
Comment (rdamazio) [reply src/model/storage/gdrive.ts:410]: Passing FormData to .request() works too, but the API library does both authentication and error handling that I'd rather not do by hand (and that, honestly, many projects I've seen use fetch() probably do wrong :) ).

File: src/model/storage/gdrive.ts [hunk 10]
```diff
@@ -0,0 +1,481 @@
...
+      return;
+    }
+
+    this._syncInterval = window.setInterval(() => {
+      // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+      if (GoogleDriveStorage.DEBUG_SYNC) console.log(`SYNC TIMER: needed=${this._needsSync}`);
+
+      // Synchronize if there are local changes, or if it's been a while (to pull remote changes).
+      const now = new Date();
+      if (this._needsSync || now.valueOf() - this._lastSync.valueOf() > GoogleDriveStorage.LONG_SYNC_INTERVAL_MS) {
```
[the remark this answers is not in the document]
Comment (rdamazio) [reply src/model/storage/gdrive.ts:448]: I had done that originally, but someone added prettier for formatting code on the project :) and prettier just removes those parenthesis as unnecessary. Extracted a variable instead.

File: src/model/storage/gdrive.ts [hunk 11]
```diff
@@ -0,0 +1,481 @@
...
+        this._lastSync = now;
```
[the remark this answers is not in the document]
Comment (rdamazio) [reply src/model/storage/gdrive.ts:449]: Yup, done.

File: src/model/storage/multipart.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (rdamazio) [reply src/model/storage/multipart.ts:1]: Yeah, I'm fine with keeping the multipart library for now. Most people I've seen using this API were actually just manually building the body as a single string, which seemed fragile.

**Comment by rdamazio** [comment]
Thanks for the review, and for calling me out on poorly-documented code (it's easy to overlook that when I'm the only one maintaining it :) ). Hopefully addressed most comments and will push again in a min.

File: src/model/storage/gdrive.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (zyv) [reply src/model/storage/gdrive.ts]: Ha-ha, sorry 😊 Yeah, that's what I would have done 👍

File: src/model/storage/gdrive.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (zyv) [reply src/model/storage/gdrive.ts]: Great idea, I like it 👍

File: src/model/storage/gdrive.ts (hunk 1 above)
Comment (zyv) [reply src/model/storage/gdrive.ts]: Thanks for the explanation, I only started to get it towards the end of the review... makes sense.

**Comment by zyv** [comment]
> Thanks for the review, and for calling me out on poorly-documented code (it's easy to overlook that when I'm the only one maintaining it :) ). 

Speaking of documentation, your latest changes to the README are really nice. However, it occurred to me that you might also want to update the boxes in the welcome component? I don't think you've done this yet.

**Comment by zyv** [comment]
> Hopefully addressed most comments and will push again in a min.

Thanks for adding the documentation and cleaning up the module, it really helps to understand the code! I'd still need more time to feel comfortable signing a statement in blood that there are no logical problems, but after another look I see no obvious issues. And I hope that such a declaration is not what you require :)

It seems I can't resolve the conversations myself, but I'm happy with this PR 👍 Just consider my comment about the boxes above.

**Comment by rdamazio** [comment]
All done, and split/folded some commits to make this easier to follow. Thanks again!

**Review by zyv — APPROVED** [review]
I had to think for a moment what would happen if you connect to Google Drive, don't like it, disconnect and delete all data (on the Drive) and then later change your mind. But you have a condition that local file mtimes are checked against mtimes of trashed files and in this case local file will be the same or newer, so they should actually be restored on the Drive 😅

I like how commits are now telling a story 🚀

**Comment by rdamazio** [comment]
> I had to think for a moment what would happen if you connect to Google Drive, don't like it, disconnect and delete all data (on the Drive) and then later change your mind. But you have a condition that local file mtimes are checked against mtimes of trashed files and in this case local file will be the same or newer, so they should actually be restored on the Drive 😅

Well, if you choose to delete all data when disconnecting, it's actually deleted, not just trashed - this is for privacy purposes (the user is asking to have their data gone, so we make it really gone). In that case, connecting again will just upload as it did the first time.
Same if you manually delete the data without disconnecting.

If you disconnect WITHOUT deleting the data, and then reconnect, you're correct that mtime will be used to re-synchronize.

**Pull request merged**